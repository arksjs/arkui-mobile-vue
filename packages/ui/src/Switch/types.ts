import type { FormItemCommonProps } from '../Form/types'

export interface SwitchProps extends FormItemCommonProps {
  modelValue?: boolean
  color?: string
  activeColor?: string
  size?: number | string
}

export interface SwitchEmits {
  onChange?: (checked: boolean) => void
}
