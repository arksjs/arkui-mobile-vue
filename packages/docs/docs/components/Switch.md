# Switch 开关选择器

注：

- 支持表单，具体可参考 [Form](./Form.md)。

<CodeDemo name="Switch">

<<< @/../../demo/src/components/Form/Switch/index.vue

</CodeDemo>

## Import

```js
import { TaSwitch } from 'tantalum-ui-mobile'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

## Props

| 属性         | 类型             | 默认值 | 必填 | 说明                            |
| ------------ | ---------------- | ------ | ---- | ------------------------------- |
| name         | string           |        | 否   | 标识                            |
| disabled     | boolean          | false  | 否   | 是否禁用                        |
| v-model      | boolean          | false  | 否   | 是否开启                        |
| color        | string           |        | 否   | 自定义默认态的背景色            |
| active-color | string           |        | 否   | 自定义激活态的背景色            |
| size         | string \| number |        | 否   | 自定义组件大小（高度），单位 px |

## Events

| 事件   | 描述               | 回调函数参数   |
| ------ | ------------------ | -------------- |
| change | 点击切换状态时触发 | value: boolean |
