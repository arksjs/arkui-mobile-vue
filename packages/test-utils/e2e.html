<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
