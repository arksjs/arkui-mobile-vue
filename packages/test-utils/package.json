{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
