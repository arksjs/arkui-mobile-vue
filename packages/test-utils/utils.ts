export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
