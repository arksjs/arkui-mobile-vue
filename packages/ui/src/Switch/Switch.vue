<template>
  <label :class="classes" :style="styles">
    <input
      class="ta-switch_checkbox"
      type="checkbox"
      :disabled="disabled"
      :name="name"
      :value="checked.toString()"
      @change="onChange"
      ref="inputEl"
    />
  </label>
</template>

<script lang="ts">
import { onMounted, ref, watch, defineComponent, computed } from 'vue'
import { formItemProps } from '../Form/form'
import { useInput } from '../Form/use-form'
import { isBoolean, colorValidator, type PropsToEmits } from '../helpers'
import { getClasses, getStyles } from './util'
import type { SwitchEmits } from './types'

export default defineComponent({
  name: 'ta-switch',
  props: {
    ...formItemProps,
    modelValue: {
      type: Boolean,
      default: null // 首次默认为null
    },
    color: {
      type: String,
      validator: colorValidator
    },
    activeColor: {
      type: String,
      validator: colorValidator
    },
    size: {
      type: [Number, String]
    }
  },
  emits: {
    'update:modelValue': value => isBoolean(value),
    change: value => isBoolean(value)
  } as PropsToEmits<SwitchEmits>,
  setup(props, { emit }) {
    let isValueNull = props.modelValue == null
    const checked = ref(false)
    const { inputEl, setInputChecked, getInputChecked } = useInput()

    const classes = computed(() => getClasses(props.disabled))
    const styles = computed(() => getStyles(props))

    function onChange() {
      const value = getInputChecked()

      checked.value = value

      if (props.modelValue !== value) {
        emit('update:modelValue', value)
      }

      emit('change', value)
    }

    onMounted(() => {
      // 同步Mounted之前设置的值到 input 元素
      setInputChecked(checked.value)
    })

    watch(
      () => props.modelValue,
      val => {
        if (isBoolean(val) && val !== checked.value) {
          checked.value = val

          setInputChecked(val)
          // 外面设置进来的，只要符合，不需要调用onChange
        } else if (isValueNull) {
          isValueNull = false
          // 首次外部没有设置值
          emit('update:modelValue', checked.value)
          emit('change', checked.value)
        }
      },
      { immediate: true }
    )

    return {
      inputEl,
      checked,
      onChange,
      classes,
      styles
    }
  }
})
</script>
