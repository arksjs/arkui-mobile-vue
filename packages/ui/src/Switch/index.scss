@import '../style/var.scss';

.ak-switch {
  --ak-color: #{$border-color};
  --ak-active-color: #{$primary-color};
  --ak-size: 32px;

  height: var(--ak-size);

  &_checkbox {
    position: relative;
    width: 1.875em;
    height: 1em;
    background: var(--ak-color);
    border: 0;
    border-radius: 1em;
    font-size: var(--ak-size);
    outline: 0;
    cursor: pointer;
    transition: all 0.2s linear;
    box-sizing: border-box;
    padding: 0;
    margin: 0;
    appearance: none;

    &:checked {
      background-color: var(--ak-active-color);
    }

    &::after {
      position: absolute;
      top: 2px;
      left: 2px;
      width: calc(100% / 1.875 - 4px);
      height: calc(100% - 4px);
      background: #fff;
      border-radius: 100%;
      transition: all 0.2s linear;
      box-sizing: border-box;
      content: '';
      box-shadow: 0 4px 4px rgba(#000, 0.12), 0 1px 2px rgba(#000, 0.02);
    }

    &:checked::after {
      left: calc(100% / 1.875 * 0.875 + 2px);
    }

    &:disabled {
      cursor: not-allowed;
      opacity: 0.2;
    }
  }
}
