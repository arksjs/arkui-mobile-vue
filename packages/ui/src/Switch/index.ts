import { withInstall } from '../helpers/with-install'
import Switch from './Switch.vue'

export { Switch }
export const install = withInstall(Switch)
export default Switch
