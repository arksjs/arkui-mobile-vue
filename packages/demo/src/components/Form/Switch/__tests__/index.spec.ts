import { shallowMount } from '@vue/test-utils'
import { Switch } from '@/Switch'

describe('Switch', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(Switch, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop disabled=true correctly', () => {
    const wrapper = shallowMount(Switch, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop value=true correctly', () => {
    const wrapper = shallowMount(Switch, {
      props: {
        modelValue: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // test('should formValue=true with input trigger change correctly', async () => {
  //   const wrapper = shallowMount(Switch, {})
  //   const input = wrapper.find('input[type="checkbox"]')

  //   await input.setChecked()

  //   expect(wrapper.vm.formValue).toBeTruthy()
  // })
})
