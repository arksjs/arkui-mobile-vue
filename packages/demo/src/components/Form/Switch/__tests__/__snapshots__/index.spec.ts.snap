// Vitest Snapshot v1

exports[`Switch > should render default correctly 1`] = `"<label class=\\"fx-switch\\"><input class=\\"fx-switch_checkbox\\" type=\\"checkbox\\" name=\\"\\" value=\\"false\\"></label>"`;

exports[`Switch > should render prop disabled=true correctly 1`] = `"<label class=\\"fx-switch disabled\\"><input class=\\"fx-switch_checkbox\\" type=\\"checkbox\\" disabled=\\"\\" name=\\"\\" value=\\"false\\"></label>"`;

exports[`Switch > should render prop value=true correctly 1`] = `"<label class=\\"fx-switch\\"><input class=\\"fx-switch_checkbox\\" type=\\"checkbox\\" name=\\"\\" value=\\"true\\"></label>"`;
